### PR TITLE
[Tune] Restore searcher and scheduler properly on `Tuner.restore`

### LINKED ISSUE #30861: [Tune] `Tuner.restore` doesn't restore searchers properly
### What happened + What you expected to happen

1. I start a Tune experiment specifying a searcher such as `OptunaSearch()`.
2. The experiment gets interrupted in the middle - many trials still need configs to be suggested.
3. I use `Tuner.restore` to resume the experiment, but the default `BasicVariantGenerator` gets used instead of the searcher, and no searcher state gets loaded back. This is because it's looking for a BasicVariantGenerator checkpoint file on this check in `TrialRunner.resume`:
```python
        if self._search_alg.has_checkpoint(self._local_checkpoint_dir):
            self._search_alg.restore_from_dir(self._local_checkpoint_dir)
```
4.  Only the existing trials with configs continue to completion, so the restored experiment may not run `num_samples` as initially specified.

### Versions / Dependencies

master

### Reproduction script

1. Run optuna example and interrupt the script in the middle with ctrl+C
```
python python/ray/tune/example/optuna_example.py
```
2. Restore in a separate script
```python
from ray import tune
# replace this with the experiment directory from the run
tuner = tune.Tuner.restore("~/ray_results/easy_objective_2022-12-01_10-08-25")
tuner.fit()
```
3. The restored run will continue with the default `BasicVariantGenerator` searcher. Only existing trials with suggested configs will continue - no new trials will run, apart from the ones already running.

On experiment interrupt:
```
Number of trials: 28/100 (4 RUNNING, 24 TERMINATED)
```
On restore:
```
Number of trials: 28 (3 PENDING, 1 RUNNING, 24 TERMINATED)
```

### Issue Severity

None

### PR DESCRIPTION
**Pull request by justinvyu**
Fixes `Tuner.restore` to use the correct searcher and scheduler type and load their checkpointed state rather than using the default `BasicVariantGenerator` search algorithm/`FIFOScheduler` scheduler.

## Why are these changes needed?

Currently, custom searchers passed in through `TuneConfig` do not get restored properly. Regardless of what searcher used in the initial experiment, `Tuner.restore` will resume the experiment with a default `BasicVariantGenerator` search algorithm, which may cause the experiment to not run a total of `num_samples` trials if any trial configs haven't been generated yet. Although experiment checkpointing does save searcher state, that state doesn't actually get loaded on restore due to the search algorithm type being mismatched. The same thing happens for schedulers.

This PR:
1. Passes in the saved searcher and scheduler that was pickled at the start of the experiment along with Tuner state, as arguments to `tune.run`.
2. Updates save/restore of searchers to track their entire state with `__dict__` rather than specifying a subset of properties.
    - TL;DR: Restore now loads the full searcher state into a "blank" searcher and continue successfully without needing to call other methods such as `searcher.set_searcher_properties`.
    ```python
    searcher = OptunaSearch()
    searcher.restore("checkpoint.pkl")
    # this restores the metric, mode, trial mapping of the saved searcher
    ```
    - This change is due to a problem I ran into when trying to restore searcher state from `Tuner.restore()`: `searcher.set_search_properties(metric, mode, config, ...)` wasn't being called since `Tuner.restore` doesn't pass in the param_space again (so the metric/mode never got set). (See [here](https://github.com/ray-project/ray/blob/6c10f3c7b983fc21c3df51be2c3ad0d2ffe8bea9/python/ray/tune/tune.py#L597).) Then, save/restore for some algorithms didn't keep track of the `metric` and `mode` attributes, so they remained `None`.
    - Many searchers already do this, so it's now more consistent across the board.
3. Improves unit tests of searcher save/restore to actually test the error above.

## Related issue number

Closes https://github.com/ray-project/ray/issues/30861
Partially addresses https://github.com/ray-project/ray/issues/30838 without including scheduler save/restore (this needs more investigation)

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by Yard1 — APPROVED** [review]
Thanks!

File: python/ray/tune/search/optuna/optuna_search.py [hunk 1]
```diff
@@ -602,34 +602,23 @@ def add_evaluated_point(
...
             save_object = pickle.load(inputFile)
-        if len(save_object) == 5:
-            (
-                self._sampler,
-                self._ot_trials,
-                self._ot_study,
-                self._points_to_evaluate,
-                self._evaluated_rewards,
-            ) = save_object
+        if isinstance(save_object, dict):
```
Comment (justinvyu) [inline python/ray/tune/search/optuna/optuna_search.py:612]: I've kept one level of backwards compatibility. For example, optuna searcher load was checking for length 5 state tuple vs. length 4. We don't really expect users to restore searchers across versions - would it be fine to remove these backwards compatibility cases fully?

File: python/ray/tune/tests/test_searchers.py [hunk 2]
```diff
@@ -520,32 +521,42 @@ def setUpClass(cls):
...
-            "2", {self.metric_name: 1, "config/a": 1.0, "time_total_s": 1}
-        )
+        self._on_trial_callbacks(searcher, "2")
         searcher.suggest("3")
-        searcher.on_trial_complete(
-            "3", {self.metric_name: 1, "config/a": 1.0, "time_total_s": 1}
-        )
+        self._on_trial_callbacks(searcher, "3")
+
+        # NOTE: Trial "not_completed" that was suggested before saving never completes
```
Comment (justinvyu) [inline python/ray/tune/tests/test_searchers.py:554]: Previously, `BOHB` and `HyperOptSearch` did not load their `_live_trial_mapping` state, so trials suggested before the restore would report results/complete without actually updating the underlying searcher (aka they wouldn't add more data to the bayesopt).

File: python/ray/tune/search/optuna/optuna_search.py (hunk 1 above)
Comment (krfricke) [reply python/ray/tune/search/optuna/optuna_search.py:612]: Let's mark this for removal (either as a comment or in a tracker sheet). I don't think it's a hard requirement but let's keep it for now!

**Review by krfricke — COMMENTED** [review]
Looks good to me! Can we add a test that uses Tuner.restore() and ensures scheduler/searcher are restored correctly?

File: python/ray/tune/tests/test_tuner_restore.py [hunk 3]
```diff
@@ -746,6 +748,56 @@ def test_restore_from_relative_path(ray_start_4_cpus, chdir_tmpdir):
...
+
+    del tuner
+    fail_marker.remove(ignore_errors=True)
+
+    tuner = Tuner.restore(str(tmpdir / "exp_name"), resume_errored=True)
+    tuner.fit()
+    searcher = tuner._local_tuner._tune_config.search_alg
+    scheduler = tuner._local_tuner._tune_config.scheduler
+    # Searcher state should get loaded correctly
+    # Total of 3 reported results (1 from before failure, 2 after restore)
```
Comment (krfricke) [inline python/ray/tune/tests/test_tuner_restore.py:792]: Nit (imo introducing these assertions makes the test a bit more readable)

```suggestion
    searcher = tuner._local_tuner._tune_config.search_alg
    scheduler = tuner._local_tuner._tune_config.scheduler
    assert isinstance(searcher, MockSearcher)
    assert isinstance(scheduler, MockScheduler)
    # Searcher state should get loaded correctly
    # Total of 3 reported results (1 from before failure, 2 after restore)
```

**Review by krfricke — APPROVED** [review]
Great!

**Comment by justinvyu** [comment]
The flaky tracker shows that the `test_tuner_restore` has timed out a few times in the past. With the added test, it's now timing out more consistently - I was planning on bumping the size to large.

**Pull request merged**